tests/functional/test_s3.py

```diff
@@ -60,62 +61,168 @@ def setUp(self):
-                    'server': 'AmazonS3',
-                },
                 'HTTPStatusCode': 200,
             }
         }
+
+        self.stubber.add_response(
+            method='head_object', service_response=head_response,
+            expected_params=self.copy_source)
+
+    def add_copy_object(self):
         copy_response = {
             'CopyObjectResult': {
-                'ETag': '"example0etag"'
+                'ETag': self.etag
             },
             'ResponseMetadata': {
                 'HTTPStatusCode': 200
             }
         }
+        expected_params = {
+            "Bucket": self.bucket,
+            "Key": self.key,
+            "CopySource": self.copy_source
+        }
         self.stubber.add_response(
-            method='head_object', service_response=head_response)
+            method='copy_object', service_response=copy_response,
+            expected_params=expected_params)
+
+    def add_copy_part(self, part_number, copy_range):
+        copy_part_response = {
+            "CopyPartResult": {
+                "ETag": self.etag
+            },
+            'ResponseMetadata': {
+                'HTTPStatusCode': 200
+            }
+        }
+        expected_params = {
+            "Bucket": self.bucket,
+            "Key": self.key,
+            "CopySource": self.copy_source,
+            "UploadId": self.upload_id,
+            "PartNumber": part_number,
+            "CopySourceRange": copy_range
+        }
         self.stubber.add_response(
-            method='copy_object', service_response=copy_response)
+            method='upload_part_copy', service_response=copy_part_response,
+            expected_params=expected_params)
 
     def test_client_copy(self):
+        self.add_single_part_copy()
         with self.stubber:
             response = self.s3.meta.client.copy(
-                self.copy_source, 'mybucket', 'mykey')
+                self.copy_source, self.bucket, self.key)
         # The response will be none on a successful transfer.
         self.assertIsNone(response)
 
     def test_bucket_copy(self):
-        bucket = self.s3.Bucket('mybucket')
+        self.add_single_part_copy()
+        bucket = self.s3.Bucket(self.bucket)
         with self.stubber:
-            response = bucket.copy(self.copy_source, 'mykey')
+            response = bucket.copy(self.copy_source, self.key)
         # The response will be none on a successful transfer.
         self.assertIsNone(response)
 
     def test_object_copy(self):
-        obj = self.s3.Object('mybucket', 'mykey')
+        self.add_single_part_copy()
+        obj = self.s3.Object(self.bucket, self.key)
         with self.stubber:
             response = obj.copy(self.copy_source)
         self.assertIsNone(response)
 
+    def test_copy_progress(self):
+        self.add_multipart_copy(5, 11)
+        transfer_config = TransferConfig(
+            multipart_chunksize=5, multipart_threshold=1)
```

Comment:
You should set the max_concurrency to 1 or the test may fail from time to time due to multiple threads making requests you cannot guarantee that it will happen in order.